lib/sgtn-client/api/translation.rb

```diff
@@ -181,27 +170,57 @@ def self.single_load(component, locale)
       end
     end
 
-    @@refresh_threads_lock = Mutex.new
-    @@refresh_threads = {}
+
+    @@refresh_semaphores = {}
+    @@refresh_semaphores_lock = Mutex.new
+    def self.get_semaphore(cache_key)
+      semaphore = @@refresh_semaphores[cache_key]
+      if semaphore.nil?
+        @@refresh_semaphores_lock.synchronize do 
+          if @@refresh_semaphores[cache_key].nil?
+            semaphore = Concurrent::Semaphore.new(100)
+            @@refresh_semaphores[cache_key] = semaphore
+          end
+        end
+      end
+      semaphore
+    end
+
     def self.single_refresh(component, locale)
       cache_key = SgtnClient::CacheUtil.get_cachekey(component, locale)
-      @@refresh_threads_lock.synchronize do
-        thread = @@refresh_threads[cache_key]
-        if thread.nil? || thread.alive? == false
-          expired, items = SgtnClient::CacheUtil.get_cache(cache_key)
-          if expired || items.nil?
-            thread = Thread.new do
```

Comment:
Oh, I just realized that the @@refresh_threads_lock.synchronize is useless since you are creating a new thread here. The threads that you create here do not lock on anything. Therefore, they will collide--it is not thread-safe, correct?

Reply:
Is this why you have locking mechanism in load_and_compare_source? I have not fully reviewed this part, but if yes, we can avoid this.

Reply:
>         thread = @@refresh_threads[cache_key]
>         if thread.nil? || thread.alive? == false
Check the code, there will be only thread to update a bundle.
I have tested it with hundreds of threads and it works well.